### PR TITLE
[WIP] :sparkles: MachinePool API (provider agnostic scale types)

### PR DESCRIPTION
**Pull request by juan-lee**
Signed-off-by: Juan-Lee Pang <jpang@microsoft.com>

**What this PR does / why we need it**:
Proposed implementation of MachinePool (provider agnostic scale types)
[CAEP](https://docs.google.com/document/d/17oyJsUYWq75by5xIMT3KyEHm0Ln__DNZp9b8AK0jszg/edit).

**Which issue(s) this PR fixes**:
Fixes #TODO

### CODE REVIEW

File: api/v1alpha2/machinepool_types.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
+/*
```
Comment (juan-lee) [inline api/v1alpha2/machinepool_types.go]: @vincepri @detiber controller-gen was giving me problems if the type didn't already exist in v1alpha2. Any idea what to do here?

**Comment by juan-lee** [comment]
Test provider that implements MachinePool can be found [here](https://github.com/juan-lee/capz). Once capz cuts a v1alpha3 branch, I'll add a [WIP] PR there that corresponds with this PR.

File: api/v1alpha2/machinepool_types.go (hunk 1 above)
Comment (vincepri) [reply api/v1alpha2/machinepool_types.go]: What was the error message? We should definitely not add the type to v1alpha2

Comment (ncdc) [reply api/v1alpha2/machinepool_types.go]: Try adding this to the v1a3 type: `+k8s:conversion-gen=false`?

**Comment by ncdc** [comment]
@juan-lee thanks for this! Before we go any further with this, we'd like to have the CAEP for MachinePools opened, reviewed, and merged first. Does that sound ok?

**Comment by juan-lee** [comment]
> @juan-lee thanks for this! Before we go any further with this, we'd like to have the CAEP for MachinePools opened, reviewed, and merged first. Does that sound ok?

@ncdc Absolutely, I'll port the changes over from the google doc when I get some cycles. I just wanted to get this in front of people as early as possible.

**Comment by ncdc** [comment]
@juan-lee should we close this now that you have #1929 ?

**Comment by vincepri** [comment]
wrong one

**Comment by juan-lee** [comment]
Closing in favor of: https://github.com/kubernetes-sigs/cluster-api/pull/1952

**Pull request closed without merging**